contrib/scalar/scalar.c

```diff
@@ -19,55 +19,75 @@ static int is_unattended(void) {
 	return git_env_bool("Scalar_UNATTENDED", 0);
 }
 
+/**
+ * Remove the deepest subdirectory in the provided path string. Path must not
+ * include a trailing path separator.
+ */
+static void strbuf_parentdir(struct strbuf *buf)
+{
+	char *path_sep = find_last_dir_sep(buf->buf);
+	strbuf_setlen(buf, path_sep ? path_sep - buf->buf : 0);
+}
+
+/**
+ * Given an absolute path at or below the enlistment root, find the
+ * enlistment root and working directory
+ */
+static void find_enlistment(struct strbuf *path,
+				   struct strbuf *enlistment_root)
+{
+	strbuf_addstr(path, "/src");
+	for (;;) {
+		size_t len = path->len;
+
+		strbuf_addstr(path, "/.git");
+		if (is_git_directory(path->buf)) {
```

Comment:
According to the issue linked to this pull request (and what I saw of the .NET Scalar implementation), the enlistment root _could_ be the git working directory itself (mostly used when `scalar register`-ing an existing `git clone`. @derrickstolee would probably be the best one to clarify on that, though.

For what it's worth, I _think_ I caught any instances where that assumption was encoded - since most operations are performed on the working directory anyway, there wasn't a lot to change.

Reply:
I think we're still missing the edge case where `<root>/src/.git` is the repository, but the current directory is `<root>/out/`.

So I think we need the following strategy:

- Loop
  - if `src/.git` is a Git directory, we are in the enlistment root
  - if `.git` is a Git directory,
    - if the current directory is named `src`, the enlistment root is the parent directory
    - otherwise the current directory is the enlistment root
  - otherwise, continue with the parent directory
  - if there is no parent directory (use `offset_1st_component()` to accommodate for DOS drives and network shares), error out

Of course, we could be a bit smart by determining the parent directory already before the `.git` check (because we need it both to test whether the current directory is called `src` as well as in the case where we continue with the parent directory).

Reply:
In addition to the changes made earlier to handle parallel directories (like `<root>/out`), this should now handle DOS/network paths with root-level scalar registrations (using `offset_1st_component`. That logic is in a separate `fixup!` commit if you'd like to review it on its own - I _think_ it avoids off-by-one errors, but I would appreciate a second opinion.